include/boost/math/statistics/linear_regression.hpp

```diff
@@ -46,35 +50,39 @@ auto simple_ordinary_least_squares(RandomAccessContainer const & x,
     return std::make_pair(c0, c1);
 }
 
-template<class RandomAccessContainer>
-auto simple_ordinary_least_squares_with_R_squared(RandomAccessContainer const & x,
-                                   RandomAccessContainer const & y)
+template<typename RandomAccessContainer, typename Real = typename RandomAccessContainer::value_type, 
+         typename Size = typename RandomAccessContainer::size_type>
+auto simple_ordinary_least_squares_with_R_squared(RandomAccessContainer const & x, RandomAccessContainer const & y)
+    -> std::tuple<Real, Real, Real>
 {
-    using Real = typename RandomAccessContainer::value_type;
-    if (x.size() <= 1)
+    if (std::size(x) <= 1)
     {
         throw std::domain_error("At least 2 samples are required to perform a linear regression.");
     }
 
-    if (x.size() != y.size())
+    if (std::size(x) != std::size(y))
     {
         throw std::domain_error("The same number of samples must be in the independent and dependent variable.");
     }
-    auto [mu_x, mu_y, cov_xy] = boost::math::statistics::means_and_covariance(x, y);
 
-    auto var_x = boost::math::statistics::variance(x);
+    const std::tuple<Real, Real, Real> temp = boost::math::statistics::means_and_covariance(x, y);
+    const Real mu_x = std::get<0>(temp);
+    const Real mu_y = std::get<1>(temp);
+    const Real cov_xy = std::get<2>(temp);
+    const Real var_x = boost::math::statistics::variance(x);
 
-    if (var_x <= 0) {
+    if (var_x <= 0) 
+    {
         throw std::domain_error("Independent variable has no variance; this breaks linear regression.");
     }
 
-
-    Real c1 = cov_xy/var_x;
-    Real c0 = mu_y - c1*mu_x;
+    // c1 is over-estimated for all types without compensation
+    const Real c1 = cov_xy/(var_x - std::numeric_limits<Real>::epsilon());
```

Comment:
I don't understand this change. Do you have a reference for this?

Could you test it with data that is all < 1e-16? (e.g., the LIGO data is all < 1e-16).

Reply:
I do not have a reference, but empirically it seems to be the case. If you multiply everything by 1e-22 (LIGO-Hanford Range) and represent with `float_quad`, `float_50`, and `float_100` the results are similarly improved.

Reply:
@mborland : Representing with `float_quad` defeats the purpose. There is no reason to use `float_quad` on LIGO-Handford data; the precision isn't there; double is fine. In this case the denominator becomes negative and the result is wrong.

I still don't understand why this improves the result.

Reply:
50% detection probability is at [3.2e-20](https://image2.slideserve.com/4246672/detection-efficiency-for-narrow-band-bursts-l.jpg) which is why I tested with quadruple and greater precision. The denominator would only go negative if `var_x` is 0 since its one epsilon of Real. The tests for variance only specify <5 epsilon and the result typically has positive error.

Reply:
3.2e-20 is fine for double precision. Double precision can represent numbers down to 1e-300.

Reply:
> The denominator would only go negative if var_x is 0 since its one epsilon of Real.

This is incorrect. You should read up on the rounding model of floating point arithmetic; see Higham, Accuracy and Stability of Numerical Algorithms. (Your statement is correct for *fixed-point* arithmetic.)

Reply:
I'll have to find a copy and give it a read. I can close this as well because it is clearly not correct.